Engine/Results/SymbolCapacity.cs

```diff
@@ -0,0 +1,279 @@
+        public bool UpdateMarketCapacity()
+        {
+            var bar = GetBar();
+            if (bar == null || bar.Volume == 0)
+            {
+                return false;
+            }
+
+            _previousBar = bar;
+
+            var utcTime = _algorithm.UtcTime;
+            var conversionRate = Security.QuoteCurrency.ConversionRate;
+            var timeBetweenBars = (decimal)(utcTime - (_previousTime ?? utcTime)).TotalMinutes;
+
+            if (_previousTime == null || timeBetweenBars == 0)
+            {
+                _previousTime = utcTime;
+                _previousVolume = bar.Volume;
+                _averageDollarVolume = conversionRate * bar.Close * bar.Volume;
+            }
+            else
+            {
+                _averageDollarVolume = ((bar.Close * conversionRate) * (bar.Volume + _previousVolume)) / timeBetweenBars;
+            }
+
+            _previousTime = utcTime;
+            _previousVolume = bar.Volume;
+
+            var includeMarketVolume = IncludeMarketVolume();
+            if (includeMarketVolume)
+            {
+                _marketCapacityDollarVolume += bar.Close * _fastTradingVolumeDiscountFactor * bar.Volume * conversionRate * Security.SymbolProperties.ContractMultiplier;
+            }
+
+            // When we've finished including market volume, signal completed
+            return !includeMarketVolume;
+        }
+
+        /// <summary>
+        /// Gets the TradeBar for the given time step. For Quotes, we convert
+        /// it into a TradeBar using market depth as a proxy for volume.
+        /// </summary>
+        /// <returns>TradeBar</returns>
+        private TradeBar GetBar()
+        {
+            TradeBar bar;
+            if (_algorithm.CurrentSlice.Bars.TryGetValue(_symbol, out bar))
+            {
+                return bar;
+            }
+
+            QuoteBar quote;
+            if (_algorithm.CurrentSlice.QuoteBars.TryGetValue(_symbol, out quote))
+            {
+                // Fake a tradebar for quote data using market depth as a proxy for volume
+                var volume = (quote.LastBidSize + quote.LastAskSize) / 2;
+                volume = _symbol.SecurityType == SecurityType.Forex
+                    ? _forexMinuteVolume
+                    : volume;
+
+                return new TradeBar(
+                    quote.Time,
+                    quote.Symbol,
+                    quote.Open,
+                    quote.High,
+                    quote.Low,
+                    quote.Close,
+                    volume);
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Signals a reset for the <see cref="MarketCapacityDollarVolume"/> and <see cref="SaleVolume"/>
+        /// </summary>
+        public void Reset()
+        {
+            _resetMarketCapacityDollarVolume = true;
```

Comment:
Why do we need the `_resetMarketCapacityDollarVolume` flag, can't we directly set to `0` the `_marketCapacityDollarVolume` here?

Reply:
ah let's also rename these `DollarVolume` all around, since it isn't true I believe because they are in account currency?

Reply:
I believe we want to maintain the capacity calculation up until a new order comes in, in case that we take a snapshot in between trades. We wouldn't want the capacity to be zero, since it'd show up as the smallest asset in the snapshot logic in `CapacityEstimate` if we reset it immediately.

I'm unable to find any terminology that diverges from Dollar Volume to describe the volume in account currency unfortunately. I'll make sure the comment addresses that point however

Reply:
We don't want to reset it immediately; we need to only reset it if there is a future trade so we can "remember" the capacity of previous trades which occurred outside the sample window